src/frontend/metrics/DeviceDiagnosticMetrics.ts

```diff
@@ -0,0 +1,198 @@
+      height: screen.height,
+      pixelRatio: PixelRatio.get(),
+    },
+    deviceType: deviceTypeToString(Device.deviceType),
+    monthlyDeviceHash: await getMonthlyHash(
+      'device diagnostics',
+      getMetricsDeviceId(),
+      new Date(),
+    ),
+  };
+
+  setIfNotNull(result, 'brand', Device.brand);
+  setIfNotNull(result, 'manufacturer', Device.manufacturer);
+  setIfNotNull(result, 'model', Device.modelId || Device.modelName);
+  setIfNotNull(
+    result,
+    'supportedCpuArchitectures',
+    Device.supportedCpuArchitectures,
+  );
+  setIfNotNull(result, 'totalMemory', Device.totalMemory);
+
+  const isEmulator = !Device.isDevice;
+  if (isEmulator) result.isEmulator = true;
+
+  return result;
+}
+
+const isAppActive = (appState: AppStateStatus): boolean =>
+  appState === 'active';
+
+const hasEnoughTimeElapsed = (): boolean => {
+  const lastSentAtMs = storage.getNumber(STORAGE_KEY) ?? -Infinity;
+  const lastSentAtDate = new Date(lastSentAtMs);
+  return (
+    !isDateValid(lastSentAtDate) ||
+    !isSameUtcMonthAndYear(lastSentAtDate, new Date())
+  );
+};
+
+const getMillisecondsUntilNextMonth = (): number => {
+  const now = new Date();
+  return beginningOfNextMonthUtc(now).valueOf() - now.valueOf();
+};
+
+export class DeviceDiagnosticMetrics {
+  #isEnabled = false;
+  #isOnline = false;
+  #isAppActive = isAppActive(AppState.currentState);
+  #isCurrentlySendingMetrics = false;
+
+  #subscriptionCleanupFns: null | (() => void)[] = null;
+
+  setEnabled(isEnabled: boolean): void {
+    this.#isEnabled = isEnabled;
+    this.#update();
+  }
+
+  async #update(): Promise<void> {
+    this.#updateSubscriptions();
+
+    const shouldSendMetrics =
+      this.#isEnabled &&
+      this.#isOnline &&
+      this.#isAppActive &&
+      !this.#isCurrentlySendingMetrics &&
+      hasEnoughTimeElapsed();
+    if (!shouldSendMetrics) return;
+
+    this.#isCurrentlySendingMetrics = true;
+
+    try {
+      await sendMetricsData({
+        ...getMetricsRequestInfo(),
+        dataToSend: await generateDeviceDiagnosticMetricsData(),
+      });
+      storage.set(STORAGE_KEY, Date.now());
+    } catch (err) {
+      Sentry.captureException(err);
+    } finally {
+      this.#isCurrentlySendingMetrics = false;
+    }
+  }
+
+  #updateSubscriptions(): void {
+    if (this.#isEnabled) {
+      if (this.#subscriptionCleanupFns) return;
+
+      const subscriptionCleanupFns: (() => void)[] = [];
+
+      subscriptionCleanupFns.push(
+        NetInfo.addEventListener(state => {
+          this.#isOnline = Boolean(state.isInternetReachable);
+          this.#update();
+        }),
+      );
+
+      const appStateSubscription = AppState.addEventListener(
+        'change',
+        appState => {
+          this.#isAppActive = isAppActive(appState);
+          this.#update();
+        },
+      );
+      subscriptionCleanupFns.push(() => appStateSubscription.remove());
+
+      const nextMonthTimeout = setLongTimeout(() => {
```

Comment:
I can't find any info online about leaving long timeouts when the app is in the background, however I feel like we should avoid doing that, in case it results in either the timeout being cleared (which will lead to unexpected behaviour), or the app being closed by the OS. I think it's safest to just run this while the app is in the foreground (active state).

Also, setLongTimeout is very clever, however if we are only handling the use-case of the app being in the foreground, the main edge-case we are trying to catch with the timeout is the user having the app open and connected over the month-end. It may be worth just removing this, and acknowledging that if a user keeps the app open for 25 days without turning off their phone or switching to another app, then metrics will fail to send, and that avoids adding additional code to maintain.

Reply:
Removed in 7deb5d846d7365b24a46cac61005bed10c2174e8.